### PR TITLE
Implement bundler mode

### PR DESCRIPTION
**Pull request by ChrisBr**
This pull request implements a cpio mode to use bundler to create an obscpio archive instead of downloading all files into the package directory and adding it as sources.

To enable it, the service needs to get called like this:
```
<services>
   <service name="bundle_gems">
     <param name="cpio">true</param>
   </service>
</services>
```

Advantages:
- spec file does not get modified with every service run
- no markers in spec file necessary (``### GEM START``)
- not used gems get automatically removed
- bundler is used to download the gems instead of harcoding URLs
- using different sources other than rubygems is supported (e.g. branches from GitHub)

If the ```cpio=true``` is not specified, the service will work as before.

By default, a global cache (over all packages) is not enabled (the old mode does it implicit by making use of the download file service). It can be setup with setting up a local rubygems cache and setting the environment variable with e.g. ```OBS_SERVICE_BUNDLE_GEMS_MIRROR_URL=localhost```. On build.o.o we will make use of https://github.com/geminabox/geminabox.

### CODE REVIEW

**Comment by ChrisBr** [comment]
@dmacvicar please review :smile: 

And could you maybe grant somebody from openSUSE/open-build-service write permissions to this repository so that we can help out maintaining it :innocent:

**Comment by jordimassaguerpla** [comment]
That is super cool. Thanks for this :) ! :+1:

**Review by dmacvicar — CHANGES_REQUESTED** [review]
I like the idea and the implementation a lot. I would only change one thing:

* I dislike "cpio" true as a boolean. This is not an "extra" option in terms of semantics, but an alternative path. I would prefer that we have a parameter `strategy` or `mode` that by default is set to eg. `spec` and that can be set to `cpio`.
Then if somebody comes with a new strategy `foo`, you will not have the issue of thinking what does `cpio: true, foo: true` means.

Then, I think a short mention in the README would be nice. Other than that :+1:

**Comment by dmacvicar** [comment]
(on a side note, I have given write access to the OBS team, software.o.o maintainers, and @jordimassaguerpla. Feel free to merge PR. Only thing: try to avoid inter-team bias, eg. if the request comes from OBS, request a review from a non-OBS developer, like @jordimassaguerpla )

File: bundle_gems [hunk 1]
```diff
@@ -58,6 +63,55 @@ unless File.exist?(gem_file_lock)
...
+        "bundle.ruby2.5 config --local mirror.https://rubygems.org #{mirror_url}"
+      )
+      logger.info stdout_and_stderr_str
+      abort(stdout_and_stderr_str) unless status.success?
+    end
+  end
+
+  Dir.chdir(options[:outdir]) do
+    stdout_and_stderr_str, status = Open3.capture2e(
+      'bundle.ruby2.5 package --no-install --path vendor/bundle'
```
Comment (lnussel) [inline bundle_gems]: ruby version hardcoded?

Comment (ChrisBr) [reply bundle_gems]: argh, that was for debugging! Will remove it!

**Comment by ChrisBr** [comment]
> I dislike "cpio" true as a boolean. This is not an "extra" option in terms of semantics, but an alternative path. I would prefer that we have a parameter strategy or mode that by default is set to eg. spec and that can be set to cpio.
Then if somebody comes with a new strategy foo, you will not have the issue of thinking what does cpio: true, foo: true means.

This is a good idea, will change it :+1:

**Comment by ChrisBr** [comment]
@dmacvicar updated your suggestions, please review again

**Review by dmacvicar — APPROVED** [review]

**Pull request merged**